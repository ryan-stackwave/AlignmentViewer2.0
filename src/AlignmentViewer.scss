/** App.scss */
@use "./MolecularStyles.module.scss";

.AlignmentViewer {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-rows: 100px 100px 40px 40px 30px 700px;
  grid-gap: 5px;
  padding: 10px;
  position: relative;
}

.testing_box {
  position: relative;
  grid-column-start: 1;
  grid-column-end: 2;
  grid-row-start: 1;
  grid-row-end: 3;
  font-size: 11px;
}

.alignment_box {
  position: relative;
  grid-column-start: 2;
  grid-column-end: 4;
  grid-row-start: 6;
  grid-row-end: 6;
}

.ace_content {
  cursor: auto;
}

/**
 * Conservation / Entropy / Gap plots
 */
.conservation_box {
  position: relative;
  grid-column-start: 2;
  grid-column-end: 4;
  grid-row-start: 1;
  grid-row-end: 1;
  /*border:solid yellow 1px;*/
}
#sequence_conservation {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  overflow-x: scroll;
}
#sequence_conservation .svg_container {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}
svg.conservation .gapCount rect {
  fill: #b7b7b7;
}
svg.conservation .entropy rect {
  fill: #000000;
}

/**
 * Logo plot
 */
.logo_box {
  position: relative;
  grid-column-start: 2;
  grid-column-end: 4;
  grid-row-start: 2;
  grid-row-end: 2;
  /*border:solid blue 1px;*/
}
.logo_box #sequence_logo {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  overflow-x: scroll;
}

/**
 * Position indicator
 */
.position_box {
  position: relative;
  grid-column-start: 2;
  grid-column-end: 4;
  grid-row-start: 5;
  grid-row-end: 5;
}
.position_box #ace-positions {
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  font-family: Monaco, Menlo, "Ubuntu Mono", Consolas, source-code-pro,
    monospace;
  font-size: 12px;
  height: 30px;
  line-height: 30px;
  padding-left: 3px;
  overflow: scroll;
  overflow-x: hidden;
  border-bottom: 1px solid red;
  border-top: 1px solid red;
}

/**
 * Query Sequence
 */
.queryseq_box {
  position: relative;
  grid-column-start: 2;
  grid-column-end: 4;
  grid-row-start: 3;
  grid-row-end: 3;
}
.queryseq_box #ace-queryseq {
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  font-family: Monaco, Menlo, "Ubuntu Mono", Consolas, source-code-pro,
    monospace;
  font-size: 12px;
  height: 40px;
  line-height: 40px;
  padding-left: 3px;
  overflow: scroll;
  overflow-x: hidden;
}

/**
 * Consensus Sequence
 */
.consensusseq_box {
  position: relative;
  grid-column-start: 2;
  grid-column-end: 4;
  grid-row-start: 4;
  grid-row-end: 4;
}
.consensusseq_box #ace-consensusseq {
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  font-family: Monaco, Menlo, "Ubuntu Mono", Consolas, source-code-pro,
    monospace;
  font-size: 12px;
  height: 40px;
  line-height: 40px;
  padding-left: 3px;
  overflow: scroll;
  overflow-x: hidden;
}

/**
 * Individual sequence data as well as the MSA as a canvas in WebGL
 */
.datatable_box {
  position: relative;
  grid-column-start: 1;
  grid-column-end: 2;
  grid-row-start: 6;
  grid-row-end: 6;
  overflow: scroll;
}

/**
 * Full ace alignment
 */
.alignment_box #ace-alignment {
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  visibility: hidden;
  border-color: gray;
  border-style: solid;
  border-width: 1px;
  z-index: 10;
}
.alignment_box #ace-alignment.loaded {
  visibility: visible;
}

/**
 * Interaction indicators
 */
#column_mouseover {
  position: absolute;
  display: none;
  pointer-events: none;
  height: 1050px;
  z-index: 1000;
  border: solid 2px purple;
}
