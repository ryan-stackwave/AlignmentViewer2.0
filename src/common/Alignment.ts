import { generateUUIDv4 } from "./Utils";
import { defineNewAlignmentMode } from "./AceAlignmentMode";
import { Nucleotide } from "./Residues";
import { SequenceSorter } from "./AlignmentSorter";
import {
  AlignmentTypes,
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
} from "./MolecularStyles";

export interface ISequence {
  id: string;
  sequence: string;
}

/**
 * Alignment
 * This class represents a multiple sequence alignment. During initialization,
 * positional statistics are generated about the alignment and a consensus
 * sequence is determined.
 *
 * The class should generally be instantiated from subclasses (e.g., FastaAlignemtn)
 * using static methods, e.g.,
 *     FastaAlignment.fromFileContents
 *     StockholmAlignment.fromFileContents
 */
export class Alignment {
  private uuid: string;
  private aceEditorMode: string | undefined;
  private name: string;
  private predictedNT: boolean;
  private sequences: Map<SequenceSorter, ISequence[]>;
  private querySequence: ISequence;
  private positionalLetterCounts = new Map<
    number,
    { [letter: string]: number }
  >();
  private globalAlphaLetterCounts: { [letter: string]: number } = {};
  private allUpperAlphaLettersInAlignmentSorted: string[];
  private consensus: {
    sequence: ISequence;
    statistics: {
      letter: string;
      position: number;
      occurrences: number;
    }[];
  };

  /**
   * Normalize all values in an object that contains all counts such that
   * the sum of those counts equals 1.
   * @param letterCounts an object that contains the same keys, but whose
   *                     values now sum to 1.
   * @param validLetters if provided, returned object will only contain letters
   *                     that are in this array.
   */
  private static normalizeLetterCounts(
    letterCounts: { [letter: string]: number },
    validLetters?: string[]
  ) {
    const totalCounts = Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc += letterCounts[letter];
      }
      return acc;
    }, 0);

    return Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc[letter] = letterCounts[letter] / totalCounts;
      }
      return acc;
    }, {} as { [letter: string]: number });
  }

  /**
   * Create a new Alignment object. The default query sequence will
   * be set to the first sequence in sequences. This can be changed
   * by calling setQuerySequence() after creation
   * @param sequences
   */
  public constructor(name: string, sequencesAsInput: ISequence[]) {
    this.uuid = generateUUIDv4();
    this.name = name;
    this.sequences = new Map<SequenceSorter, ISequence[]>();
    this.sequences.set(SequenceSorter.INPUT, sequencesAsInput);
    this.querySequence = sequencesAsInput[0];
    this.predictedNT = true;

    //
    //generate statistics
    //
    const start = new Date();
    const allLettersInAlignment: { [key: string]: number } = {}; //all letters in the alignment

    // aggregate stats for each position and globally including
    // the number of time each amino acid occurs
    // ** this loop takes the bulk of the initialization time
    for (
      let sequenceIdx = 0;
      sequenceIdx < sequencesAsInput.length;
      sequenceIdx++
    ) {
      const seq = sequencesAsInput[sequenceIdx].sequence;
      for (let positionIdx = 0; positionIdx < seq.length; positionIdx++) {
        const letter = seq[positionIdx];
        const letterIsAlpha = letter.match(/[a-z]/i) ? true : false;
        if (
          letterIsAlpha &&
          this.predictedNT &&
          Nucleotide.fromSingleLetterCode(letter) === Nucleotide.UNKNOWN
        ) {
          this.predictedNT = false;
        }

        const position = positionIdx; // zero based
        allLettersInAlignment[letter] = 1;

        let letterCounts = this.positionalLetterCounts.get(position);
        if (!letterCounts) {
          letterCounts = {};
          this.positionalLetterCounts.set(position, letterCounts);
        }
        if (letter in letterCounts === false) {
          letterCounts[letter] = 0;
        }
        letterCounts[letter] += 1;

        if (letterIsAlpha) {
          if (letter in this.globalAlphaLetterCounts === false) {
            this.globalAlphaLetterCounts[letter] = 0;
          }
          this.globalAlphaLetterCounts[letter] += 1;
        }
      }
    }

    //get a list of all the upper letters found in the alignment
    this.allUpperAlphaLettersInAlignmentSorted = Object.keys(
      allLettersInAlignment
    )
      .sort()
      .reduce((arr: string[], value: string) => {
        if (value.match(/[A-Z]/)) {
          // only keep letters
          arr.push(value);
        }
        return arr;
      }, []);

    //determine consensus sequence
    const consensusStats = Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        const topLetter = Object.keys(letterCounts).reduce((a, b) => {
          const aIsLetter = a.match(/[a-z]/i);
          const bIsLetter = b.match(/[a-z]/i);
          if (aIsLetter && !bIsLetter) {
            return a;
          }
          if (bIsLetter && !aIsLetter) {
            return b;
          }
          return letterCounts[a] > letterCounts[b] ? a : b;
        });
        return {
          position: position,
          letter: topLetter,
          occurrences: letterCounts[topLetter],
        };
      }
    );
    this.consensus = {
      sequence: {
        id: "consensus",
        sequence: consensusStats.map((s) => s.letter).join(""),
      },
      statistics: consensusStats,
    };

    console.log(
      "done parsing alignment. took " +
        (new Date().getTime() - start.getTime()) +
        "ms"
    );
  }

  /**
   * Is this alignment predicted to be nucleotides?
   * @returns true if it is predicted to be nucleotide sequences.
   */
  getPredictedType(): AlignmentTypes {
    return this.predictedNT
      ? AlignmentTypes.NUCLEOTIDE
      : AlignmentTypes.AMINOACID;
  }

  /**
   * Returns a default nucleotide or amino acid style for this alignment,
   * predicted from the sequence itself.
   */
  getDefaultStyle(): AminoAcidAlignmentStyle | NucleotideAlignmentStyle {
    return this.predictedNT
      ? new NucleotideAlignmentStyle()
      : new AminoAcidAlignmentStyle();
  }

  /**
   * Get a unique uuid for this alignment.
   * @returns a uuid string for this alignment
   */
  getUUID() {
    return this.uuid;
  }

  /**
   * Get the name of this alignment
   * @returns a name for the alignment - usually the filename
   */
  getName() {
    return this.name;
  }

  /**
   * Get an ace editor mode for this alignemnt. If one doesn't exist
   * then it will be defined.
   * @returns the ace editor mode for this alignment
   */
  getAceEditorMode() {
    if (!this.aceEditorMode) {
      this.aceEditorMode = "ace/mode/" + this.uuid;
      defineNewAlignmentMode(this.aceEditorMode, this);
    }
    return this.aceEditorMode;
  }

  /**
   * Get the consensus sequence
   * @returns a position ordered array of objects that contain each letter, its number of
   *          occurrences, and the actual position number
   */
  getConsensus() {
    return this.consensus;
  }

  /**
   * Get the number of gaps at each position
   * @returns an array of gap counts, ordered by position
   */
  getGapsPerColumn(): number[] {
    return Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        return !isNaN(letterCounts["-"]) ? letterCounts["-"] : 0;
      }
    );
  }

  /**
   * Get the number of gaps at a specific position
   * @returns the gap counts at the specified position
   */
  getGapCountAtColumn(position: number): number {
    const letterCounts = this.positionalLetterCounts.get(position);
    if (!letterCounts || !letterCounts["-"]) {
      return 0;
    }
    return letterCounts["-"];
  }

  /**
   * Get the total number of times each letter appears in the alignment.
   * @param normalize if true, will normalize the returned array such that
   *                  it sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   */
  getGlobalAlphaLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize) {
      return Alignment.normalizeLetterCounts(
        this.globalAlphaLetterCounts,
        validLetters
      );
    }
    return !validLetters
      ? this.globalAlphaLetterCounts
      : Object.keys(this.globalAlphaLetterCounts).reduce((acc, letter) => {
          //don't normalize, but remove invalid letters
          if (validLetters.includes(letter)) {
            acc[letter] = this.globalAlphaLetterCounts[letter];
          }
          return acc;
        }, {} as { [letter: string]: number });
  }

  /**
   * Get the length of the longest sequence
   * @returns the length of the largest sequence in this alignment
   */
  getMaxSequenceLength(): number {
    return this.getSequences().reduce((acc, seq) => {
      if (seq.sequence.length > acc) {
        acc = seq.sequence.length;
      }
      return acc;
    }, -1);
  }

  /**
   * Get the frequencies of all letters in this alignment for each position.
   * @param normalize if true, will normalize the returned letter counts for each
   *                  position such that each sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   * @returns a Map whose keys are position (ordered) and values are a dictionary
   *          with [key = letter (e.g., amino acid code)] and [value = # occurrences].
   * TODO: Test parameter use cases.
   */
  getPositionalLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize || validLetters) {
      return Array.from(this.positionalLetterCounts).reduce(
        (acc, [position, letterCounts]) => {
          acc.set(
            position,
            normalize
              ? Alignment.normalizeLetterCounts(letterCounts, validLetters) // normalize + remove invalid letters
              : Object.keys(letterCounts).reduce((acc2, letter) => {
                  //don't normalize, but remove invalid letters
                  if (validLetters!.includes(letter)) {
                    acc2[letter] = letterCounts[letter];
                  }
                  return acc2;
                }, {} as { [letter: string]: number })
          );
          return acc;
        },
        new Map<Number, { [letter: string]: number }>()
      );
    }
    return this.positionalLetterCounts;
  }

  /**
   * Get all the sequences in this alignment - ordered as they were input
   * @param sortBy specifies the order of the returned sequences. If it isn't
   *               provided, returns the sequences as they were input to create
   *               the Alignment object, e.g., usually fasta order.
   * @param forceReSort if provided, sequences are forced to be re-sorted (rather
   *                    than possibly cached)
   * @returns all sequences in this alignment
   */
  getSequences(sortBy?: SequenceSorter, forceReSort?: boolean): ISequence[] {
    sortBy = sortBy ? sortBy : SequenceSorter.INPUT;
    forceReSort = forceReSort ? forceReSort : false;
    if (
      sortBy !== SequenceSorter.INPUT &&
      (!this.sequences.has(sortBy) || forceReSort)
    ) {
      //cache not yet populated for this sort option
      this.sequences.set(
        sortBy,
        sortBy.sortFn(this.sequences.get(SequenceSorter.INPUT)!, this)
      );
    }
    return this.sequences.get(sortBy)!;
  }

  /**
   * Get a sorted list of all of the upper case letters in the alignment.
   * @returns a list of all letters in the alignment sorted alphabetically.
   */
  getAllUpperAlphaLettersInAlignmentSorted(): string[] {
    return this.allUpperAlphaLettersInAlignmentSorted;
  }

  /**
   * Get all the first sequence in the alignment, aka, target / query / first
   * @returns all sequences in this alignment
   */
  getQuerySequence(): ISequence {
    return this.querySequence;
  }

  /**
   * Set the query sequence
   * @param newQuery The new query sequence. Must already exist in the alignment.
   * @throws an error if the query sequence does not yet exist in the alignment
   *
   * IMPORTANT: Requires re-sorting
   */
  /*setQuerySequence(newQuery: ISequence) {
    const queryInAlignment = this.getSequences().find((seq) => {
      if (seq === newQuery) {
        return true;
      }
      return false;
    });
    if (queryInAlignment !== undefined) {
      this.querySequence = newQuery;

      //the resort could be really slow ... need to rethink.
      SequenceSorter.list.forEach((sortOption) => {
        //updated cached sorted sequences
        if (sortOption.isQuerySequenceDependent) {
          this.getSequences(sortOption, true);
        }
      });
      this.sequences.forEach(
        (sequences: ISequence[], sortBy: SequenceSorter) => {
          this.getSequences(sortBy, true); // force re-sorting, which updates cache
        }
      );
    } else {
      throw Error("query sequence does not exist in alignment");
    }
  }*/
}
