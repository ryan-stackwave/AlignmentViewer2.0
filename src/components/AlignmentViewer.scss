/** AlignmentViewer.scss */
$annotationContentRightGapSize: 20px;
$aceEditorLineHeight: 20px;

.alignment-viewer.annotation-closed .av-widget .av-annotation {
  display: none;
}

.alignment-viewer {
  display: flex;
  flex-flow: column;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;

  //single row ace editors
  .av-widget.av-consensus-seq-holder,
  .av-widget.av-query-seq-holder,
  .av-widget.av-position-indicator-holder {
    height: $aceEditorLineHeight;

    .ace_editor {
      line-height: $aceEditorLineHeight;
    }
  }

  //the main ace editor
  .av-widget.av-ace-msa-holder {
    flex-grow: 4;
  }

  .av-widget {
    display: flex;
    padding: 4px 8px;

    .av-annotation,
    .av-content {
      display: flex;
      overflow: auto;
    }
    .av-content::-webkit-scrollbar {
      background: transparent; // make scrollbar transparent
      height: 0;
    }
    .av-annotation {
      flex: 0 1 15%;
      justify-content: flex-end;
      align-items: center;
      // bit of a hack to put space between annotation and content
      margin-right: $annotationContentRightGapSize;
      font-weight: bold;
      text-align: right;
      position: relative;
    }

    .av-content {
      flex: 1 0 85%; //annotations take up 15%
      position: relative;
      align-items: center;
    }

    .ace_editor {
      position: absolute;
      top: 0px;
      left: 0px;
      right: 0px;
      bottom: 0px;
      overflow: auto;
      overflow-x: hidden;
      overflow-y: hidden;
    }
  }
}
