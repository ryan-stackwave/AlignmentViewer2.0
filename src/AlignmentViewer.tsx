import React from "react";
import "./AlignmentViewer.scss";
import { Ace } from "ace-builds";
import Alignment, { SequenceSortOptions } from "./Alignment";
import ScrollSync, { ScrollType } from "./ScrollSync";
import { AceMSAComponent, AceEditorTypes } from "./AceMSAComponent";
import { SequenceLogoComponent, LOGO_TYPES } from "./SequenceLogoComponent";
import { SequenceConservationComponent } from "./SequenceConservationComponent";
import { AlignmentCanvasComponent } from "./AlignmentCanvasComponent";
import {
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
  PositionsToStyle
} from "./MolecularStyles";

export interface AppProps {
  alignment: Alignment;
  style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle;
  logoPlotStyle: LOGO_TYPES;
  zoomLevel: number;
  sortBy: SequenceSortOptions;
}
export interface AppState {
  aceCharacterWidth: number;
  aceEditors: Ace.Editor[];
  alignmentEditorFirstRow?: number;
  alignmentEditorLastRow?: number;
}

export class AlignmentViewer extends React.Component<AppProps, AppState> {
  constructor(props: AppProps) {
    super(props);
    this.state = {
      aceEditors: [],
      aceCharacterWidth: 0
    };

    //setup scroll groups
    ScrollSync.getInstance().setScrollerGroup("horiz", ScrollType.horizontal);
  }

  _handleCharacterSizeChanged = (newCharSize: number) => {
    this.setState({
      aceCharacterWidth: newCharSize
    });
  };

  _aceEditorLoaded = (id: string, editor: Ace.Editor) => {
    console.log(
      "_aceEditorLoaded id =" + id + ", metrics:",
      //@ts-ignore
      editor.renderer.$fontMetrics
    );
    let scrollSync = ScrollSync.getInstance();
    scrollSync.registerScroller(editor, "horiz");
    this.setState({
      aceCharacterWidth: editor.renderer.characterWidth, //todo: check if the same always.
      aceEditors: [editor].concat(this.state.aceEditors)
    });

    console.log("");
    //track visible rows to show in canvas MSA
    if (id === "ace-alignment") {
      editor.renderer.on("afterRender", () => {
        // BREAKS LOGO, rerenders everything below and kills performance. React lifecycle stuff.
        setTimeout(() => {
          this.setState({
            alignmentEditorFirstRow: editor.renderer.getFirstFullyVisibleRow(),
            alignmentEditorLastRow: editor.renderer.getLastFullyVisibleRow()
          });
        });
      });
    }
  };

  _elementLoaded(id: string, scroller: HTMLElement) {
    console.log("_elementLoaded id =" + id);
    let scrollSync = ScrollSync.getInstance();
    scrollSync.registerScroller(scroller, "horiz");
  }

  render() {
    return !this.props.alignment ? null : (
      <div className="AlignmentViewer">
        <div id="column_mouseover"></div>
        {this.renderConservationBox()}
        {this.renderSequenceLogo()}
        {this.renderConsensusQueryBox()}
        {this.renderQuerySeqBox()}
        {this.renderPositionBox()}
        {this.renderDatatableBox()}
        {this.renderAlignmentBox()}
      </div>
    );
  }

  protected renderSequenceLogo = () => (
    <div
      className={`logo_box ${this.props.style.alignmentType.className}
                       ${this.props.style.colorScheme.className} 
                       ${PositionsToStyle.ALL.className}`}
    >
      {
        <SequenceLogoComponent
          id="sequence_logo"
          alignment={this.props.alignment}
          glyphWidth={this.state.aceCharacterWidth}
          logoType={this.props.logoPlotStyle}
          logoLoaded={element => {
            this._elementLoaded("sequence_logo", element as HTMLElement);
          }}
        />
      }
    </div>
  );
  protected renderConservationBox = () => (
    <div className="conservation_box">
      {
        <SequenceConservationComponent
          id="sequence_conservation"
          alignment={this.props.alignment}
          characterWidth={this.state.aceCharacterWidth}
          conservationPlotLoaded={element => {
            this._elementLoaded("sequence_conservation", element);
          }}
        />
      }
    </div>
  );

  protected renderConsensusQueryBox = () => (
    <div className="consensusseq_box">
      {
        <AceMSAComponent
          id="ace-consensusseq"
          type={AceEditorTypes.consensus}
          alignment={this.props.alignment}
          fontSize={this.props.zoomLevel}
          sortBy={this.props.sortBy}
          topLevelClassNames={[
            this.props.style.residueDetail.className,
            this.props.style.alignmentType.className,
            this.props.style.positionsToStyle.className,
            this.props.style.colorScheme.className
          ].join(" ")}
          editorLoaded={editor => {
            this._aceEditorLoaded("ace-consensusseq", editor);
          }}
        />
      }
    </div>
  );

  protected renderQuerySeqBox = () => (
    <div className="queryseq_box">
      {
        <AceMSAComponent
          id="ace-queryseq"
          type={AceEditorTypes.query}
          alignment={this.props.alignment}
          fontSize={this.props.zoomLevel}
          sortBy={this.props.sortBy}
          topLevelClassNames={[
            this.props.style.residueDetail.className,
            this.props.style.alignmentType.className,
            this.props.style.positionsToStyle.className,
            this.props.style.colorScheme.className
          ].join(" ")}
          editorLoaded={editor => {
            this._aceEditorLoaded("ace-queryseq", editor);
          }}
        />
      }
    </div>
  );

  protected renderPositionBox = () => (
    <div className="position_box">
      {
        <AceMSAComponent
          id="ace-positions"
          type={AceEditorTypes.position}
          alignment={this.props.alignment}
          fontSize={this.props.zoomLevel}
          sortBy={this.props.sortBy}
          editorLoaded={editor => {
            this._aceEditorLoaded("ace-positions", editor);
          }}
        />
      }
    </div>
  );

  protected renderDatatableBox = () => (
    <div className="datatable_box" id="datatable">
      {
        <AlignmentCanvasComponent
          id="alignment_canvas"
          alignment={this.props.alignment}
          alignmentType={this.props.style.alignmentType}
          positionsToStyle={this.props.style.positionsToStyle}
          colorScheme={this.props.style.colorScheme}
          sortBy={this.props.sortBy}
          mouseDown={(x, y) => {
            console.log("someone clicked on the canvas:" + x + "," + y);
          }}
          highlightRows={
            this.state.alignmentEditorFirstRow !== undefined &&
            this.state.alignmentEditorLastRow !== undefined
              ? [
                  this.state.alignmentEditorFirstRow,
                  this.state.alignmentEditorLastRow
                ]
              : undefined
          }
        />
      }
    </div>
  );

  protected renderAlignmentBox = () => (
    <div className="alignment_box">
      <AceMSAComponent
        id="ace-alignment"
        type={AceEditorTypes.alignment}
        alignment={this.props.alignment}
        fontSize={this.props.zoomLevel}
        sortBy={this.props.sortBy}
        topLevelClassNames={[
          this.props.style.residueDetail.className,
          this.props.style.alignmentType.className,
          this.props.style.positionsToStyle.className,
          this.props.style.colorScheme.className
        ].join(" ")}
        characterSizeChanged={this._handleCharacterSizeChanged}
        editorLoaded={editor => {
          this._aceEditorLoaded("ace-alignment", editor);
        }}
      />
    </div>
  );
}
