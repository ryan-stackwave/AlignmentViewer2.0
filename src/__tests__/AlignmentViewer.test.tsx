import * as React from "react";
import { shallow, default as Enzyme } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import "jest-canvas-mock";
import Alignment from "../Alignment";
import { AlignmentViewer } from "../AlignmentViewer";
import { AlignmentStyle, AlignmentTypes } from "../MolecularStyles";
import { LOGO_TYPES } from "../SequenceLogoComponent";

// Due to the runtime necessities of using styles, we need to explicitly mock out some stub data.
// https://github.com/facebook/jest/issues/3094
jest.mock("../MolecularStyles.module.scss", () => {
  return {
    aaStyBGAlpha_Default: "",
    aaStyClass_Default: "",
    aaStyColorOrder_Default: "",
    aaStyColors_Default: "",
    aaStyDesc_Default: "",
    ntStyBGAlpha_Default: "",
    ntStyClass_Default: "",
    ntStyColorOrder_Default: "",
    ntStyColors_Default: "",
    ntStyDesc_Default: ""
  };
});

describe("AlignmentViewer", () => {
  beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });

  it("Should render", () => {
    const wrapper = shallow(
      <AlignmentViewer
        alignment={new Alignment("", [])}
        style={AlignmentStyle.fromAlignmentType(AlignmentTypes.NUCLEOTIDE)}
        zoomLevel={0}
        logoPlotStyle={LOGO_TYPES.BARS}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
