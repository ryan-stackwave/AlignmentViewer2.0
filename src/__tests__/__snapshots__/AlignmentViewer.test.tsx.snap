// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlignmentViewer Should render 1`] = `
<div
  className="AlignmentViewer"
>
  <div
    id="column_mouseover"
  />
  <div
    className="conservation_box"
  >
    <SequenceConservationComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      characterWidth={0}
      conservationPlotLoaded={[Function]}
      id="sequence_conservation"
    />
  </div>
  <div
    className="logo_box undefined
                        
                       undefined"
  >
    <SequenceLogoComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      glyphWidth={0}
      id="sequence_logo"
      logoLoaded={[Function]}
      logoType="bar plot"
    />
  </div>
  <div
    className="consensusseq_box"
  >
    <AceMSAComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      editorLoaded={[Function]}
      fontSize={0}
      id="ace-consensusseq"
      sortBy={
        SequenceSortOptions {
          "description": "As input",
          "distanceFn": undefined,
          "isTargetSequenceDependent": false,
          "key": "as-input",
        }
      }
      topLevelClassNames="   "
      type={1}
    />
  </div>
  <div
    className="queryseq_box"
  >
    <AceMSAComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      editorLoaded={[Function]}
      fontSize={0}
      id="ace-queryseq"
      sortBy={
        SequenceSortOptions {
          "description": "As input",
          "distanceFn": undefined,
          "isTargetSequenceDependent": false,
          "key": "as-input",
        }
      }
      topLevelClassNames="   "
      type={0}
    />
  </div>
  <div
    className="position_box"
  >
    <AceMSAComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      editorLoaded={[Function]}
      fontSize={0}
      id="ace-positions"
      sortBy={
        SequenceSortOptions {
          "description": "As input",
          "distanceFn": undefined,
          "isTargetSequenceDependent": false,
          "key": "as-input",
        }
      }
      type={3}
    />
  </div>
  <div
    className="datatable_box"
    id="datatable"
  >
    <AlignmentCanvasComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      alignmentType={
        AlignmentTypes {
          "allColorSchemes": Array [
            Object {
              "backgroundAlpha": NaN,
              "className": "",
              "colors": Object {
                "": "",
              },
              "commonName": "Default",
              "description": "",
            },
          ],
          "className": undefined,
          "description": "Nucleotide Sequences",
          "key": "nucleotide-alignment",
        }
      }
      colorScheme={
        Object {
          "backgroundAlpha": NaN,
          "className": "",
          "colors": Object {
            "": "",
          },
          "commonName": "Default",
          "description": "",
        }
      }
      id="alignment_canvas"
      mouseDown={[Function]}
      positionsToStyle={
        PositionsToStyle {
          "className": undefined,
          "description": "All",
          "key": "all",
        }
      }
      sortBy={
        SequenceSortOptions {
          "description": "As input",
          "distanceFn": undefined,
          "isTargetSequenceDependent": false,
          "key": "as-input",
        }
      }
    />
  </div>
  <div
    className="alignment_box"
  >
    <AceMSAComponent
      alignment={
        Alignment {
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
        }
      }
      characterSizeChanged={[Function]}
      editorLoaded={[Function]}
      fontSize={0}
      id="ace-alignment"
      sortBy={
        SequenceSortOptions {
          "description": "As input",
          "distanceFn": undefined,
          "isTargetSequenceDependent": false,
          "key": "as-input",
        }
      }
      topLevelClassNames="   "
      type={2}
    />
  </div>
</div>
`;
